internal.nix

```diff
@@ -331,6 +353,16 @@ rec {
 
               ${preInstallLinkCommands}
 
+              # patchShebangs will only patch executable files
+              if [[ "$(pwd)" != "/build" ]]; then # ignore the root package. bin entries only make sense for dependencies
+                jq -r 'select(.bin != null) | .bin | if type == "string" then . else values[] end' package.json | while read binTarget; do
+                  if [[ ! -x "$binTarget" ]]; then
+                    echo "make binary executable: $(pwd)/$binTarget"
+                    chmod +x "$binTarget" || exit 1 # on error, throw ELIFECYCLE
+                  fi
+                done
+              fi
```

Comment:
@milahu At least this part of this PR relies on hook scripts, see the `node_modules` `buildPhase`

Reply:
aah yes, thanks
so this is part of the `node_modules/.hooks/prepare` script
which makes this PR useless for npmv7

which brings me back to https://github.com/nix-community/npmlock2nix/issues/110#issuecomment-921263996
&rarr; lets implement our own `npm install`
instead of producing workarounds for a npm-without-hooks

Reply:
> I think being able to patch sources is useful on its own

yes, of course. like `patchPhase` in `stdenv.mkDerivation`

for example, `npm install puppeteer` will download a full chromium binary
to `node_modules/puppeteer/.local-chromium/`

in nix, we can say

```nix
{
  pname = "puppeteer";
  patchPhase = ''
    <package.json jq '.scripts.install = ""' | sponge package.json
  '';
  postInstall = ''
    d=$out/node_modules/puppeteer/.local-chromium/linux-970485
    mkdir -p $d
    ln -s ${puppeteer-chromium} $d/chrome-linux
  '';
```

related: [patching json files in place](https://stackoverflow.com/questions/36565295/jq-to-replace-text-directly-on-file-like-sed-i)

> Implementing an entirely separate `npm install` feels a bit over the top to me, requiring maintenance, the potential for bugs, having to synchronize versions, adding a lot of complexity.

a working prototype is implemented in my [pnpm-install-only](https://github.com/milahu/pnpm-install-only)

the complexity is limited by

1. focussing only on `npm install`. its not rocket science:
   * verify tar.gz files
   * unpack tar.gz files
   * run lifecycle scripts
   * populate `node_modules/.bin`
3. using an external library ([snyk-nodejs-lockfile-parser](https://github.com/snyk/nodejs-lockfile-parser)) to parse lockfiles. bonus: this supports different lockfile formats: npm-shrinkwrap.json package-lock.json yarn.lock. pnpm-lock.yaml is [supported only via workaround](https://github.com/snyk/nodejs-lockfile-parser/issues/111), but it "should" be easy to implement. (seems everybody is a frontend dev these days ...)
4. using `npm run` to run lifecycle scripts. no need to reimplement that part of npm

another bonus: this custom `npm install` allows to use the nix store as a machine-level store for node packages, see my [nodejs-hide-symlinks](https://github.com/milahu/nodejs-hide-symlinks)

also that project is a working prototype. its just not popular, because of ... reasons. more human reasons than technical reasons